### PR TITLE
re-fix filtering unindexed messages in squat backend

### LINKED ISSUE #4692: Not yet squattered emails are always returned on any search (3.4 and later)
Hi, I've been digging this problem using gdb but I hardly can figure out what's happening here.

Using squatter mode in imapd.conf, when recent emails arrive and are not yet been squattered, any search will return the correct result of the squattered emails + all of these recent emails, which are not filtered considering the query.
This can be easily simulated by commenting out the rolling squatter from cyrus.conf, then sending a couple of new emails to the test account: try to do queries and you will always receive these new emails as a result, even if they don't match the query expression.

I discovered that the problem is during subquery_run_indexed() (search_query.c), after running the squatter query via bx->run() it will run the subquery_post_enginesearch callback for any folder. Here it will run index_search_evaluate(), where at the beginning runs search_expr_always_same: this returns 1 because e->op is SEOP_TRUE, causing the code to not consider the original query and going on adding the uids as a match.

This search expression is sent to index_search_evaluate from sub->expr, which is the one containing SEOP_TRUE, while I've seen previous code checking the "op" inside sub->indexed, I don't know what this means actually.
I also can't find why this expr is SEOP_TRUE when the search is a SEOP_MATCH (justs a search "FROM abcd"), and infact sub->indexed reflects it.

@rsto @elliefm you may already have checked something in the past with me about it, but maybe with these new considerations we may find where the bug is?

Here is  the previous issue: #4224 

Gabriele

### PR DESCRIPTION
**Pull request by elliefm**
This was originally fixed in #4698, except it turns out there was additional complexity required that was missed because our tests were inadequate.

The search is split into indexed and unindexed subexpressions.  The indexed subexpression is run through the search engine backend, and the unindexed subexpression is then applied as a filter over the returned results.

The squat engine returns all unindexed messages in its matched set, expecting the post-processing phase to filter them further.  This has been broken (#4692) for a while, because nowadays the post-processing phase doesn't apply the indexed subexpression that squat expects it to, only the unindexed subexpression.

The original fix in #4698 failed to recognise the search having previously been split into separate subexpressions.  It copied the indexed subexpression over the unindexed subexpression and used that for filtering, which had the desired effect when the whole search could be completed from the search index, but broke if there were also unindexed terms (e.g. `SINCE date`) to filter on which had now been lost.

This new attempt at a fix is similar, but recombines the indexed and unindexed subexpressions instead of replacing one with the other.  It also adds a bunch of tests!

Fixes #4692

### CODE REVIEW

**Comment by elliefm** [comment]
@ksmurchison Can I get a review on this please?  I'm blocked on some things until this lands.  Thanks

**Review by ksmurchison — APPROVED** [review]

**Pull request merged**
